src/backend/cdb/motion/ic_udpifc.c

```diff
@@ -1167,6 +1169,107 @@ resetRxThreadError()
 }
 
 
+/*
+ * isIPv6Available
+ * 		Check whether a valid IPv6 address exists in the system.
+ *
+ * 		We want to have both loopback IPv6 address, which will show if IPv6 was
+ * 		disabled earlier, and some non-link-local address currently available
+ * 		for our use.
+ */
+static bool
+isIPv6Available(void)
+{
+	bool		hasIpv6Addr = false;
+	bool		hasIpv6Loopback = false;
+	int			errnoSave = errno;
+
+	int			errorErrnoSave = 0;
+	char	   *errorCause = NULL;
+
+	struct ifaddrs	*ifaddr = NULL;
+	struct ifaddrs	*ifaddrFirst = NULL;
+
+	if (getifaddrs(&ifaddrFirst) == -1)
+	{
+		errorCause = "getifaddrs()";
+		goto end;
+	}
+
+	for (ifaddr = ifaddrFirst; ifaddr != NULL; ifaddr = ifaddr->ifa_next)
+	{
+		if (ifaddr->ifa_addr != NULL &&
+			ifaddr->ifa_addr->sa_family == AF_INET6)
+		{
+			const struct sockaddr_in6 *addr =
+				(struct sockaddr_in6 *) ifaddr->ifa_addr;
+			const struct in6_addr *sinAddr = &addr->sin6_addr;
+
+			int			bindFd = -1;
+			int			checkResult = -1;
+
+			/* Discard wildcard and link-local addresses. */
+			if (IN6_IS_ADDR_UNSPECIFIED(sinAddr) ||
+				IN6_IS_ADDR_LINKLOCAL(sinAddr))
+			{
+				continue;
+			}
+
+			/* Make sure the address can be used in bind(). */
+			bindFd = socket(AF_INET6, SOCK_DGRAM, 0);
+
+			if (bindFd < 0)
+			{
+				if (errno != EAFNOSUPPORT)
+				{
+					errorCause = "socket()";
+					goto end;
+				}
+
+				break;
+			}
+
+			checkResult = bind(bindFd, (struct sockaddr *) addr,
+							   sizeof(struct sockaddr_in6));
+
+			closesocket(bindFd);
+
+			if (checkResult == 0)
+			{
+				if (IN6_IS_ADDR_LOOPBACK(sinAddr))
+					hasIpv6Loopback = true;
+				else
+					hasIpv6Addr = true;
+
+				if (hasIpv6Addr && hasIpv6Loopback)
+					break;
+			}
+		}
+	}
+
+end:
+	errorErrnoSave = errno;
+
+	if (ifaddrFirst != NULL)
+		freeifaddrs(ifaddrFirst);
+
+	if (errorCause != NULL)
+	{
+		errno = errorErrnoSave;
```

Comment:
Do we really need the `errorErrnoSave`? Do you expect that `freeifaddrs` can modify `errno`?

Reply:
`free()` may change `errno` due to `munmap()` failure before glibc 2.33.

Source:
https://sourceware.org/pipermail/libc-announce/2021/000030.html (2021-02-01):
> The following bugs are resolved with this release:
>
>  ...
>  [17924] malloc: 'free' should not set errno